R/mediate.R

```diff
@@ -0,0 +1,386 @@
+               mediator = NULL, assay.type = NULL, dim.type = NULL,
+               family = gaussian(), covariates = NULL, p.adj.method = "BH",
+               add.metadata = FALSE, message = TRUE, ...) {
+
+        ###################### Input check ########################
+        if (!outcome %in% names(colData(se))) {
+          stop(outcome, " not found in colData(se).", call. = FALSE)
+        }
+        if (!treatment %in% names(colData(se))) {
+          stop(treatment, " not found in colData(se).", call. = FALSE)
+        }
+        if (!is.null(covariates) & !all(covariates %in% names(colData(se)))) {
+          stop("covariates not found in colData(se).", call. = FALSE)
+        }
+        if (!.is_a_bool(add.metadata)) {
+          stop("add.metadata must be TRUE or FALSE.", call. = FALSE)
+        }
+        if (!.is_a_bool(message)) {
+          stop("message must be TRUE or FALSE.", call. = FALSE)
+        }
+        
+        # Check that arguments can be passed to mediate and remove unused samples
+        se <- check.mediate.args(se, outcome, treatment, mediator, covariates, ...)
+        
+        # Check which mediator was provided (colData, assay or reducedDim)
+        med_opts <- sapply(
+          list(mediator = mediator, assay.type = assay.type, dim.type = dim.type),
+          function(x) !is.null(x)
+        )
+        
+        if (sum(med_opts) == 1) {
+          if (med_opts[[1]]) {
+            # Check that mediator is in colData
+            if (!mediator %in% names(colData(se))) {
+              stop(mediator, " not found in colData(se).", call. = FALSE)
+            }
+            # Use mediator for analysis  
+            mediators <- mediator
+            mat <- NULL
+                
+          } else if (med_opts[[2]]) {
+            # Check that assay is in assays
+            if (!assay.type %in% assayNames(se)) {
+              stop(assay.type, " not found in assays(se).", call. = FALSE)
+            }
+            # Define matrix for analysis
+            mat <- assay(se, assay.type)
+            # Use assay for analysis
+            mediators <- rownames(mat)
+                
+          } else if (med_opts[[3]]) {
+            # Check that reducedDim is in reducedDims
+            if (!dim.type %in% reducedDimNames(se)) {
+              stop(dim.type, " not found in reducedDims(se).", call. = FALSE)
+            }
+            # Define matrix for analysis
+            mat <- t(reducedDim(se, dim.type))
+            # Give component names to matrix rows
+            rownames(mat) <- paste0(dim.type, seq(1, nrow(mat)))
+            # Use reducedDim for analysis
+            mediators <- rownames(mat)
+          }
+          
+        } else {
+          # Throw error if none or multiple mediation options are specified
+          stop("The arguments mediator, assay.type and dim.type are mutually exclusive",
+               ", but ", sum(med_opts), " were provided.", call. = FALSE)
+        }
```

Comment:
I moved it up to the first part of the conditional. However, I don't think there's a way to completely remove the if else statement, since input is different for assay, reducedDims and colData.